index.js

```diff
@@ -348,42 +295,81 @@ function asThrift(opts, request, onResponse) {
     try {
         sender.send(request, opts.endpoint, opts.head,
             opts.body, onResponse);
-    } catch(e) {
-        if (e.message ===
-            fmt('type %s_args not found', opts.endpoint)) {
-            var emsg = fmt('%s endpoint does not exist', opts.endpoint);
-            onResponse(new Error(emsg));
+    } catch (err) {
+        // TODO untangle this mess
+        if (err.message === fmt('type %s_args not found', opts.endpoint)) {
+            delegate.error(fmt('%s endpoint does not exist', opts.endpoint));
+            done();
+            return delegate.exit();
         } else {
-            var msg = e.message || '';
-            reportError(opts.logger,
-                'Error response received for the as-thrift request. '
-                + msg, e);
+            delegate.error('Error response received for the as-thrift request.');
+            delegate.error(err);
+            done();
+            return delegate.exit();
         }
     }
-}
 
-function asRaw(opts, request, onResponse) {
-    request.headers.as = 'raw';
+    function onResponse(err, res, arg2, arg3) {
+        done();
+        self.onResponse(err, res, arg2, arg3, opts, delegate);
+    }
+};
+
+TCurl.prototype.asRaw = function asRaw(opts, request, delegate, done) {
+    var self = this;
+    request.headers.as = opts.as || 'raw';
     request.send(opts.endpoint, opts.head, opts.body,
         onResponse);
-}
 
-function asHTTP(opts, client, subChan, onResponse, logger) {
-    var ashttp = TCurlAsHttp({
+    function onResponse(err, res, arg2, arg3) {
+        done();
+        self.onResponse(err, res, arg2, arg3, opts, delegate);
+    }
+};
+
+TCurl.prototype.asHTTP = function asHTTP(opts, client, subChan, delegate, done) {
+    var asHttp = TCurlAsHttp({
         channel: client,
         subChannel: subChan,
         method: opts.http,
         endpoint: opts.endpoint,
         headers: opts.head,
         body: opts.body,
-        onResponse: onResponse,
-        logger: logger
+        done: done,
+        logger: delegate
     });
-    ashttp.send();
-}
+    asHttp.send();
+};
 
-function asJSON(opts, request, onResponse) {
+TCurl.prototype.asJSON = function asJSON(opts, request, delegate, done) {
+    var self = this;
     var sender = new TChannelAsJSON();
     sender.send(request, opts.endpoint, opts.head,
         opts.body, onResponse);
+
+    function onResponse(err, res, arg2, arg3) {
+        done();
+        self.onResponse(err, res, arg2, arg3, opts, delegate);
+    }
+};
+
+TCurl.prototype.onResponse = function onResponse(err, res, arg2, arg3, opts, delegate) {
+    if (err) {
+        delegate.error(err);
+        return delegate.exit();
```

Comment:
The nice thing about using the delegate object instead of a callback is that there isn’t necessarily one error(), so I’m making these orthogonal.